### PR TITLE
Enable DE runs by filtering out annotation labels with only a single observation (SCP-4282)

### PR DESCRIPTION
**Pull request by jlchang**
Scanpy's DE function will exit with a ValueError ("Could not calculate statistics for groups {annotation_label} since they only contain one sample.") and no DE results are generated if ANY annotation_label has only one observation. This PR checks the reconstituted data object undergoing DE analysis for annotation labels that have only one observation (ie. cell/bead) and deletes the data that would trigger the problematic error.

To test, run DE using a test metadata file(de_singlesample_metadata.tsv) where seurat_cluster "6" would have a single sample "Mm_AMB_N107":

activate the scp-ingest-pipeline repo virtualenv
then from the scripts directory of the scp-ingest-pipeline repo, perform this setup:

```
source ../scripts/setup_mongo_dev.sh <path to your Github token file>
unset BARD_HOST_URL
```

Run the DE job from the ingest directory of the scp-ingest-pipeline repo:
`python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 differential_expression --annotation-name seurat_clusters --annotation-type group --annotation-scope cluster --matrix-file-path ../tests/data/differential_expression/de_integration.tsv --matrix-file-type dense --annotation-file ../tests/data/differential_expression/de_singlesample_metadata.tsv --cluster-file ../tests/data/differential_expression/de_integration_annotated_cluster.txt --cluster-name de_singlesample --study-accession SCPdev --differential-expression`

confirm that the job runs successfully and output is generated for seurat_clusters 0-4

> de_singlesample--seurat_clusters--0--cluster--wilcoxon.tsv
> de_singlesample--seurat_clusters--1--cluster--wilcoxon.tsv
> de_singlesample--seurat_clusters--2--cluster--wilcoxon.tsv
> de_singlesample--seurat_clusters--3--cluster--wilcoxon.tsv
> de_singlesample--seurat_clusters--4--cluster--wilcoxon.tsv

This satisfies SCP-4282.

### CODE REVIEW

**Review by eweitz — APPROVED** [review]
Code looks good!  I note a few nits, nothing blocking.

File: tests/test_de.py [hunk 1]
```diff
@@ -86,6 +87,43 @@ def test_assess_annotation(self):
...
+            number_removed,
+            1,
+            f"expected removal of single sample, found {number_removed}",
+        )
+
+        removed = list(set(adata.obs_names) - set(adata2.obs_names))
+        self.assertEqual(
+            removed,
+            ["Mm_AMB_N107"],
+            f"expected removal of cell [\'Mm_AMB_N107\'], not {removed}",
```
[the remark this answers is not in the document]
Comment (jlchang) [reply tests/test_de.py]: `6b7b1e6`

File: ingest/de.py [hunk 2]
```diff
@@ -229,12 +228,12 @@ def get_genes(genes_path):
...
             # print so we're aware of dups during dev testing
-            if genes_df[1].count() == genes_df[1].nunique():
+            if genes_df[1].count() != genes_df[1].nunique():
                 msg = "dev_info: Features file contains duplicate identifiers (col 2)"
                 print(msg)
             return genes_df[1].tolist()
         else:
-            if genes_df[0].count() == genes_df[0].nunique():
+            if genes_df[0].count() != genes_df[0].nunique():
                 msg = "dev_info: Features file contains duplicate identifiers (col 1)"
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/de.py]: `6b7b1e6`

File: ingest/de.py [hunk 3]
```diff
@@ -229,12 +228,12 @@ def get_genes(genes_path):
         if len(genes_df.columns) > 1:
             # unclear if falling back to gene_id is useful (SCP-4283)
             # print so we're aware of dups during dev testing
-            if genes_df[1].count() == genes_df[1].nunique():
+            if genes_df[1].count() != genes_df[1].nunique():
                 msg = "dev_info: Features file contains duplicate identifiers (col 2)"
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/de.py]: `6b7b1e6`

**Review by devonbush — APPROVED** [review]
nice

**Pull request merged**